
<!DOCTYPE html>
<html>
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <title>Template Alignment: 7cj3.1 | Report</title>
    <meta name="description" content="SWISS-MODEL modelling results for Template Alignment: 7cj3.1." />
    <link rel="stylesheet" type="text/css" href="../web/js/jquery.fancybox.css"/>
    <style>
			.sequence { font-size: 12px; font-family:"Courier New", Courier, monospace;}
			h2 {font-size: 1.2em; font-family:Arial; margin:20 0 0 0}
			h3 {font-size: 1.0em; font-family:Arial; margin:0 0 0 0}
			
			.borderedTbl>tbody>tr>td, .borderedTbl th{ 
        text-align:center;
  			border-top:1px solid #DDD;
  			border-left:1px solid #DDD;
      }
      .borderedTbl{
        border-bottom:1px solid #DDD;
        border-right:1px solid #DDD; 
      }
      td { 
        -webkit-hyphens: auto;
           -moz-hyphens: auto;
                hyphens: auto; 
              }

			body {
			width:100%; 
			height:100%; 
			background-color: #E7E8EA;
			margin: 0;
			padding: 0;
			word-wrap: break-word;
			font-family: "Helvetica Neue",Helvetica,Arial,sans-serif;
      font-size: 14px;

			}
			#content{
			padding:20px;
			width:780px;
			margin-left:auto;
			margin-right:auto;
			background-color:#FFF;}
			
			#sib_footer {
			background-color: #333333;
			color: #F7F8FA;
			font-family: Verdana;
			font-size: 10px;
			font-weight: normal;
			height: 20px;
			letter-spacing: 1px;
			
			width:760px;
      margin-left:auto;
      margin-right:auto;
      padding:0px 20px 0px 20px
			}
			
			#sib_footer a {
  			color: #FFFFFF;
  			font-family: Verdana;
  			font-size: 10px;
  			font-weight: normal;
  			letter-spacing: 1px;
			}
			a, a:link, a:visited {
			color: #D6161D;
			text-decoration: none;
			} 	

			.alTbl{ font-family:'Courier New', Courier, monospace; }
	   img{ border:0; }
			
	@media print{
	
	#content{
			padding:20px;
			background-color:#FFF;
		}
	body{ 
	    width:100%;
	    height:100%; 
	    background-color: #FFF;
			padding: 0;
		}	
	#sib_footer{
		display:none;
	}
	}
      
    </style>
  </head>
  <body>
  
  <div id="content">
  	<table border=0 width="100%">
  		<tr>
    		<td valign="middle">
  			<a title="SIB Swiss Institute of Bioinformatics" href="https://www.sib.swiss">
  			<img src="../web/images/sib_logo_small.png"></a>
  		</td>
  		<td valign="middle">
  			<a title="Biozentrum, University of Basel" href="https://www.biozentrum.unibas.ch">
  			<img src="../web/images/biozentrum_en.png"></a>
  		</td>
		  <td valign="middle"><div style="font-size:1.9em;">SWISS-MODEL Homology Modelling Report</div> </td>
	 </tr>
	</table> 
    
  <div style="clear:both"></div>


 
  <h2>Model Building Report</h2>
  <p>
This document lists the results for the homology modelling project &quot;Template Alignment: 7cj3.1&quot; submitted to SWISS-MODEL workspace
on July 28, 2021, 4:01 p.m..The submitted primary amino acid sequence is given in Table T1.</p>
<p>If you use any results in your research, please cite the relevant publications:</p>
<ul>
  <li>Waterhouse, A., Bertoni, M., Bienert, S., Studer, G., Tauriello, G., Gumienny, R.,
      Heer, F.T., de Beer, T.A.P., Rempfer, C., Bordoli, L., Lepore, R., Schwede, T.   
      SWISS-MODEL: homology modelling of protein structures and complexes.
      Nucleic Acids Res. 46(W1), W296-W303 (2018).
    <a href="https://pubmed.ncbi.nlm.nih.gov/29788355"
       target="_blank"><img src="../web/images/pubmed.png
                                 "
                            width="17" height="16" align="absmiddle"></a>
    <a href="https://dx.doi.org/10.1093/nar/gky427"
       target="_blank"><img src="../web/images/doi.jpg
                                 "
                            width="30" height="16" align="absmiddle"
                            style="padding-left:2px"></a>
  </li>
  <li>Bienert, S., Waterhouse, A., de Beer, T.A.P., Tauriello, G., Studer,
    G., Bordoli, L., Schwede, T. The SWISS-MODEL Repository - new features and
    functionality. Nucleic Acids Res. 45, D313-D319 (2017).
    <a href="https://pubmed.ncbi.nlm.nih.gov/27899672"
       target="_blank"><img src="../web/images/pubmed.png
                                 "
                            width="17" height="16" align="absmiddle"></a>
    <a href="https://dx.doi.org/10.1093/nar/gkw1132"
       target="_blank"><img src="../web/images/doi.jpg
                                 "
                            width="30" height="16" align="absmiddle"
                            style="padding-left:2px"></a>
  </li>
  <li><a name="promod3"></a>Studer, G., Tauriello, G., Bienert, S.,     
          Biasini, M., Johner, N., Schwede, T. ProMod3 - A versatile homology 
          modelling toolbox. PLOS Comp. Biol. 17(1), e1008667 (2021).
    <a href="https://pubmed.ncbi.nlm.nih.gov/33507980"
       target="_blank"><img src="../web/images/pubmed.png
                                 "
                            width="17" height="16" align="absmiddle"></a>
    <a href=" https://doi.org/10.1371/journal.pcbi.1008667"
       target="_blank"><img src="../web/images/doi.jpg
                                 "
                            width="30" height="16" align="absmiddle"
                            style="padding-left:2px"></a>
  </li>
  <li><a name="qmean"></a>Studer, G., Rempfer, C., Waterhouse, A.M., 
    Gumienny, G., Haas, J., Schwede, T. QMEANDisCo - distance constraints 
    applied on model quality estimation. Bioinformatics 36, 1765-1771 (2020).
    <a href="https://pubmed.ncbi.nlm.nih.gov/31697312"
       target="_blank"><img src="../web/images/pubmed.png
                                 "
                            width="17" height="16" align="absmiddle"></a>
    <a href="https://dx.doi.org/10.1093/bioinformatics/btz828"
       target="_blank"><img src="../web/images/doi.jpg
                                 "
                            width="30" height="16" align="absmiddle"
                            style="padding-left:2px"></a>
  </li>
  <li><a name="qsqe"></a>Bertoni, M., Kiefer, F., Biasini, M., Bordoli, L.,
    Schwede, T. Modeling protein quaternary structure of homo- and
    hetero-oligomers beyond binary interactions by homology. Scientific
    Reports 7 (2017).
    <a href="https://pubmed.ncbi.nlm.nih.gov/28874689"
       target="_blank"><img src="../web/images/pubmed.png
                                 "
                            width="17" height="16" align="absmiddle"></a>
    <a href="https://dx.doi.org/10.1038/s41598-017-09654-8"
       target="_blank"><img src="../web/images/doi.jpg
                                 "
                            width="30" height="16" align="absmiddle"
                            style="padding-left:2px"></a>
  </li>
</ul>

<h2>Results</h2>
<p>
The SWISS-MODEL template library  was searched 

 
 

 for evolutionary related structures matching the target sequence in Table T1. For details on the template search, see Materials and Methods. Overall 1 templates were found (Table T2).</p>


<h2>Models</h2>


<p>The following model was built (see Materials and Methods "Model Building"):</p>

      

      <table id="model_01" class="borderedTbl" width="100%" border=0 cellspacing=1 cellpadding=5>
  		<thead>
        <tr>
        <th><h3>Model #01</h3></th>
        <th>File</th>
        <th>Built with</th>
        <th>Oligo-State</th>
        <th>Ligands</th>
        <th title="Global Model Quality Estimate">GMQE</th>
        <th>QMEANDisCo Global</th>
        </tr>
      </thead>
      <tbody>
      <tr>
        <td><a href="models/01/01.png" target="_blank"><img height="60" src="models/01/01.png"></a>
        </td>
  	    <td><a href="models/01/model.pdb" target="_blank">PDB</a></td>
  		<td>ProMod3 3.2.0</td>
        <td>monomer (requested by user)</td>
        <td>
          <div style="max-width:350px;">None
          </div>
        </td>
        <td>0.44</td>
        <td><span id="avgLocalScore_01">0.56</span> &plusmn; 0.05</td>
      </tr>
     </table>
  	
  	<table width="100%" cellspacing=5>
  	<tr>
  	    <td width="30%">
       	 <table id="qmeanGlobalTbl01"></table>
        </td>
          <td width="33%">
            <a href="#lq-chart-01" rel="lq-01" class="fancybox">
              <div id="lq-chart-01-thumb" style="margin:12px 4px;"></div>
            </a>
            <div id="lq-chart-01" style="display:none"></div>
          </td>

          <td width="33%">
            <a href="#qn-chart-01" class="fancybox">
              <div id="qn-chart-01-thumb" style="margin:12px 4px;"></div>
            </a>
            <div id="qn-chart-01" style="display:none"></div>
          </td>
  	 </tr>
  	</table>


  	<table class="borderedTbl">
  		<thead>
  			<tr>
  				<th>Template</th>
  				<th>Seq Identity</th>
  				<th>Oligo-state</th>
  				<th>QSQE</th>
  				<th>Found by</th>
  				<th>Method</th>
  				<th>Resolution</th>
  				<th>Seq Similarity</th>
  				<th>Range</th>
  				<th>Coverage</th>
  				<th>Description</th>
  			</tr>
  		</thead>
  		<tbody>
        
        
        <tr>
  				<td>
                
              <a href="models/01/templates/7cj3.1.A.pdb.gz" target="_blank">7cj3.1.A</a>
                
              
          </td>
  				<td>23.14</td>
  				<td>homo-dimer</td>
  				<td>-</td>
  				<td>USER ALIGNMENT</td>
  				<td>X-ray</td>
  				<td>2.60&Aring;</td>
  				<td>0.31</td>
  				<td>26 - 278

        </td>
  				<td>0.87</td>
  				<td>
          Phosphodiesterase</td>
  			</tr>
  		</tbody>
  	</table> 
  	
    

    

  
    <br>
    <h3>Excluded ligands</h3>
    <table width="100%" class="borderedTbl">
      <thead>
        <tr>
          <th>Ligand Name.Number</th>
          <th>Reason for Exclusion</th>
          <th>Description</th>
        </tr>
      </thead>
    <tbody>
    
      
      <tr><td>OLC.2</td><td>
      Binding site not conserved.
      </td>
      <td><div style="max-width:650px;">(2R)-2,3-dihydroxypropyl (9Z)-octadec-9-enoate</div></td>
      </tr>
      
      <tr><td>OLC.3</td><td>
      Binding site not conserved.
      </td>
      <td><div style="max-width:650px;">(2R)-2,3-dihydroxypropyl (9Z)-octadec-9-enoate</div></td>
      </tr>
      
      <tr><td>OLC.4</td><td>
      Binding site not conserved.
      </td>
      <td><div style="max-width:650px;">(2R)-2,3-dihydroxypropyl (9Z)-octadec-9-enoate</div></td>
      </tr>
      
      <tr><td>OLC.5</td><td>
      Binding site not conserved.
      </td>
      <td><div style="max-width:650px;">(2R)-2,3-dihydroxypropyl (9Z)-octadec-9-enoate</div></td>
      </tr>
      
      <tr><td>OLC.6</td><td>
      Binding site not conserved.
      </td>
      <td><div style="max-width:650px;">(2R)-2,3-dihydroxypropyl (9Z)-octadec-9-enoate</div></td>
      </tr>
      
      <tr><td>OLC.7</td><td>
      Binding site not conserved.
      </td>
      <td><div style="max-width:650px;">(2R)-2,3-dihydroxypropyl (9Z)-octadec-9-enoate</div></td>
      </tr>
      
      <tr><td>OLC.8</td><td>
      Binding site not conserved.
      </td>
      <td><div style="max-width:650px;">(2R)-2,3-dihydroxypropyl (9Z)-octadec-9-enoate</div></td>
      </tr>
      
      <tr><td>OLC.10</td><td>
      Binding site not conserved.
      </td>
      <td><div style="max-width:650px;">(2R)-2,3-dihydroxypropyl (9Z)-octadec-9-enoate</div></td>
      </tr>
      
      <tr><td>OLC.11</td><td>
      Binding site not conserved.
      </td>
      <td><div style="max-width:650px;">(2R)-2,3-dihydroxypropyl (9Z)-octadec-9-enoate</div></td>
      </tr>
      
      <tr><td>OLC.12</td><td>
      Binding site not conserved.
      </td>
      <td><div style="max-width:650px;">(2R)-2,3-dihydroxypropyl (9Z)-octadec-9-enoate</div></td>
      </tr>
      
      <tr><td>OLC.13</td><td>
      Binding site not conserved.
      </td>
      <td><div style="max-width:650px;">(2R)-2,3-dihydroxypropyl (9Z)-octadec-9-enoate</div></td>
      </tr>
      
      <tr><td>OLC.14</td><td>
      Not in contact with model.
      </td>
      <td><div style="max-width:650px;">(2R)-2,3-dihydroxypropyl (9Z)-octadec-9-enoate</div></td>
      </tr>
      
      <tr><td>OLC.15</td><td>
      Binding site not conserved.
      </td>
      <td><div style="max-width:650px;">(2R)-2,3-dihydroxypropyl (9Z)-octadec-9-enoate</div></td>
      </tr>
      
      <tr><td>OLC.16</td><td>
      Binding site not conserved.
      </td>
      <td><div style="max-width:650px;">(2R)-2,3-dihydroxypropyl (9Z)-octadec-9-enoate</div></td>
      </tr>
      
    
      
      <tr><td>RET.1</td><td>
      Binding site not conserved.
      </td>
      <td><div style="max-width:650px;">RETINAL</div></td>
      </tr>
      
      <tr><td>RET.9</td><td>
      Binding site not conserved.
      </td>
      <td><div style="max-width:650px;">RETINAL</div></td>
      </tr>
      
    
		</tbody>
	</table>
  

<br>


<pre>Target    -----------MSALTKARDALANTSAKDPYHEAK-PMGFVWLNFGMWLAAVGYTIHQRLFGGYHTK----AVDRGNSDF<br>7cj3.1.A  GRGKRRAKTRN---------------------IAIASTKEVQWQGIFMIIVWLCVMGSLIFFANPEASRRVFAKFSHLQS<br><br>Target    IYKGTTFVFIVAVCFTYLGYIGTKSAAKKSMTFVLMMTCGVVATSWFLLYL-RAGITLLDGAGHPVDIIRHLEWAHDEAN<br>7cj3.1.A  FYGATSVAFAFATGLDILAYVNAVSDEKRVLSGILAYVDGVACISYLSMATLNLYFLVDSTQGNPVWLMRYAEWIITCPT<br><br>Target    LAYFLGLLTSADQWTIGRTILESHGLLLFGFLASIA-RHPYDELFATISMCI-----------YLACLQDIISMFQKAID<br>7cj3.1.A  LLYWCGLASRADRSSVSDIATADALLLAGGALSSILPSWP------------AFFVFAGSFATYIYVMLHMWGMFGKAMQ<br><br>Target    GEVDNK-VDT--WTLR--RCRDITIFAWSYITLAWYLVRHGFWSFETGELHIALGEFCAKVVLMLIFVNNSV--------<br>7cj3.1.A  PDFQPPP---PLPRHALHLLRCEIVMSWSIFPLVEFLRRQGYIDFQVGEAMNCVADYAAKVGLAMIMVNCNLEQINALRV<br><br>Target    ----<br>7cj3.1.A  ----<br><br></pre>


  <br>
  <hr><br>
  





<h2>Materials and Methods</h2>










<h2>Target / Template Alignment</h2>
The user uploaded the following target/template alignment. 
<br>
<span class="sequence">>Target<br>-----------MSALTKARDALANTSAKDPYHEAK-PMGFVWLNFGMWLAAVGYTIHQRLFGGYHTK----AVDRGNSDFIYKGTTFVFIVAVCFTYLGY<br>IGTKSAAKKSMTFVLMMTCGVVATSWFLLYL-RAGITLLDGAGHPVDIIRHLEWAHDEANLAYFLGLLTSADQWTIGRTILESHGLLLFGFLASIA-RHP<br>YDELFATISMCI-----------YLACLQDIISMFQKAIDGEVDNK-VDT--WTLR--RCRDITIFAWSYITLAWYLVRHGFWSFETGELHIALGEFCAK<br>VVLMLIFVNNSV------------<br>>7cj3.1.A Phosphodiesterase | offset=1<br>GRGKRRAKTRN---------------------IAIASTKEVQWQGIFMIIVWLCVMGSLIFFANPEASRRVFAKFSHLQSFYGATSVAFAFATGLDILAY<br>VNAVSDEKRVLSGILAYVDGVACISYLSMATLNLYFLVDSTQGNPVWLMRYAEWIITCPTLLYWCGLASRADRSSVSDIATADALLLAGGALSSILPSWP<br>------------AFFVFAGSFATYIYVMLHMWGMFGKAMQPDFQPPP---PLPRHALHLLRCEIVMSWSIFPLVEFLRRQGYIDFQVGEAMNCVADYAAK<br>VGLAMIMVNCNLEQINALRVENLY<br></span>








<h2>Model Building</h2>
<p>
Models are built based on the target-template alignment using ProMod3 (<a href="#promod3">Studer et al.</a>). Coordinates which are conserved between the target and the template are copied from the template to the model. Insertions and deletions are remodelled using a fragment library. Side chains are then rebuilt. Finally, the geometry of the resulting model is regularized by using a force field.
</p>

<h2>Model Quality Estimation</h2>
<p>
The global and per-residue model quality has been assessed using the QMEAN scoring function (<a href="#qmean">Studer et al.</a>).
</p>

<h2>Ligand Modelling</h2>
<p>
Ligands present in the template structure are transferred by homology to the model when the following criteria are met: (a) The ligands are annotated as biologically relevant in the template library, (b) the ligand is in contact with the model, (c) the ligand is not clashing with the protein, (d) the residues in contact with the ligand are conserved between the target and the template. If any of these four criteria is not satisfied, a certain ligand will not be included in the model. The model summary includes information on why and which ligand has not been included.</p>

<h2>Oligomeric State Conservation</h2>
<p>
The quaternary structure annotation of the template is used to model the target sequence in its oligomeric form. The method (<a href="#qsqe">Bertoni et al.</a>) is based on a supervised machine learning algorithm, Support Vector Machines (SVM), which combines interface conservation, structural clustering, and other template features to provide a quaternary structure quality estimate (QSQE). The QSQE score is a number between 0 and 1, reflecting the expected accuracy of the interchain contacts for a model built based a given alignment and template. Higher numbers indicate higher reliability. This complements the GMQE score which estimates the accuracy of the tertiary structure of the resulting model.</p>

<h2>References</h2>
<p>
  <ul>
    <li><a name="blastp"></a><b>BLAST</b><br>
        Camacho, C., Coulouris, G., Avagyan, V., Ma, N., Papadopoulos, J.,
        Bealer, K., Madden, T.L. BLAST+: architecture and applications. BMC
        Bioinformatics 10, 421-430 (2009).
        <a target="_blank" href="https://pubmed.ncbi.nlm.nih.gov/20003500">
          <img width="17" align="absmiddle" height="16"
               src="../web/images/pubmed.png
                    "></a>
        <a href="https://dx.doi.org/10.1186/1471-2105-10-421"
           target="_blank"><img src="../web/images/doi.jpg
                                     " width="30"
                                height="16" align="absmiddle"
                                style="padding-left:2px"></a></li>
    <li><a name="hhblits"></a><b>HHblits</b><br>
         Steinegger, M., Meier, M., Mirdita, M., V&ouml;hringer, H., Haunsberger,
        S. J., S&ouml;ding, J. HH-suite3 for fast remote homology detection and
        deep protein annotation. BMC Bioinformatics 20, 473 (2019).
        <a target="_blank" href="https://pubmed.ncbi.nlm.nih.gov/31521110">
          <img width="17" align="absmiddle" height="16"
               src="../web/images/pubmed.png
                    "></a>
        <a href="https://dx.doi.org/10.1186/s12859-019-3019-7"
           target="_blank"><img src="../web/images/doi.jpg
                                     " width="30"
                                height="16" align="absmiddle"
                                style="padding-left:2px"></a></li>
    
  </ul>

<h2>Table T1:</h2>
<p>Primary amino acid sequence for which templates were searched and models were built.</p>
<div class="sequence">MSALTKARDALANTSAKDPYHEAKPMGFVWLNFGMWLAAVGYTIHQRLFGGYHTKAVDRGNSDFIYKGTTFVFIVAVCFTYLGYIGTKSAAKKSMTFVLM<br>MTCGVVATSWFLLYLRAGITLLDGAGHPVDIIRHLEWAHDEANLAYFLGLLTSADQWTIGRTILESHGLLLFGFLASIARHPYDELFATISMCIYLACLQ<br>DIISMFQKAIDGEVDNKVDTWTLRRCRDITIFAWSYITLAWYLVRHGFWSFETGELHIALGEFCAKVVLMLIFVNNSV<br></div>


<h2>Table T2:</h2>
<table class="borderedTbl">
<thead>
	<tr>
		<th>Template</th>
		<th>Seq Identity</th>
		<th>Oligo-state</th>
		
		<th>Found by</th>
		<th>Method</th>
		<th>Resolution</th>
		<th>Seq Similarity</th>
		<th>Coverage</th>
		<th>Description</th>
	</tr>
</thead>
<tbody>

<tr>
		<td>7cj3.1.A</td>
		<td>23.14</td>
		<td>homo-dimer</td>
		
		<td>USER ALIGNMENT</td>
		<td>X-ray</td>
		<td>2.60&Aring;</td>
		<td>0.31</td>
		<td>0.87</td>
		<td>Phosphodiesterase</td>
	</tr>
</tbody>
</table>

 


	</div>
	
  <div id="sib_footer">			
   <div id="sib_footer_content"> 
       <a href="https://www.sib.swiss">Swiss Institute of Bioinformatics</a>
       <a href="mailto:help-swissmodel@unibas.ch" style="margin-left:16px;">Contact Us</a>
      <div id="sib_footer_right" style="float:right"></div>
  	</div>
	</div>
  
  <script src="../web/js/svgtocanvas.min.js"></script>
  <script src="../web/js/jquery-3.4.1.min.js"></script>
  <script src="../web/js/jquery.fancybox.pack.js"></script> 
  <script src="../web/js/qmean.min.js"></script>
  <script src="../web/js/raphael-min.js"></script>
  <script>
 var localQMEANNormImage='../web/images/qmean_norm.png';
 var modelData = {
   
      "01":{
        'qmean_local' : { 
        "A":[null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,0.2,0.24,0.28,0.35,0.33,0.39,0.46,0.43,0.61,0.52,0.42,0.51,0.54,0.47,0.51,0.58,0.33,0.34,0.33,0.33,0.3,0.28,0.49,0.43,0.52,0.49,0.31,0.27,0.3,0.31,0.4,0.4,0.43,0.38,0.5,0.55,0.6,0.51,0.51,0.6,0.57,0.56,0.56,0.61,0.62,0.58,0.6,0.65,0.63,0.58,0.64,0.62,0.62,0.58,0.61,0.56,0.57,0.6,0.52,0.54,0.42,0.47,0.52,0.52,0.54,0.53,0.52,0.5,0.52,0.57,0.6,0.56,0.65,0.64,0.61,0.57,0.64,0.68,0.67,0.63,0.67,0.67,0.61,0.62,0.56,0.55,0.53,0.52,0.38,0.41,0.3,0.42,0.37,0.43,0.44,0.45,0.45,0.4,0.43,0.48,0.53,0.54,0.57,0.58,0.56,0.59,0.62,0.63,0.63,0.63,0.66,0.67,0.71,0.61,0.69,0.71,0.71,0.63,0.7,0.75,0.62,0.64,0.67,0.69,0.63,0.62,0.64,0.63,0.66,0.62,0.6,0.55,0.69,0.7,0.71,0.63,0.72,0.7,0.66,0.65,0.71,0.62,0.69,0.68,0.68,0.63,0.63,0.66,0.58,0.6,0.62,0.56,0.5,0.55,0.45,0.48,0.53,0.54,0.56,0.6,0.6,0.6,0.49,0.61,0.56,0.48,0.48,0.66,0.66,0.66,0.68,0.72,0.76,0.73,0.71,0.73,0.73,0.71,0.73,0.62,0.59,0.61,0.6,0.6,0.53,0.49,0.41,0.36,0.42,0.39,0.36,0.39,0.48,0.48,0.38,0.4,0.33,0.31,0.32,0.5,0.65,0.58,0.64,0.67,0.72,0.7,0.67,0.73,0.64,0.67,0.64,0.68,0.67,0.66,0.71,0.56,0.62,0.66,0.64,0.42,0.36,0.57,0.61,0.57,0.63,0.57,0.62,0.66,0.68,0.61,0.65,0.62,0.68,0.71,0.67,0.72,0.66,0.64,0.67,0.7,0.66,0.66,0.64,0.63,0.55,0.6,0.59,0.42,0.51,0.44,0.44,0.45,0.33] },
        'qmean_global':{"qmean4_norm_score":0.5944856687,"qmean4_z_score":-4.4141645325,"qmean6_norm_score":0.5977919964,"qmean6_z_score":-3.8976819323,"interaction_norm_score":-0.0186030601,"interaction_z_score":-1.1007656698,"cbeta_norm_score":0.0021732477,"cbeta_z_score":-3.8727923071,"packing_norm_score":-0.361402614,"packing_z_score":-0.2025975641,"torsion_norm_score":0.1131450004,"torsion_z_score":-3.8509768336,"ss_agreement_norm_score":0.6412411778,"ss_agreement_z_score":-0.0771186376,"acc_agreement_norm_score":0.5889328063,"acc_agreement_z_score":-0.8500275053,"avg_local_score":0.5596824683,"avg_local_score_error":0.052}
      }
    
 }  


 $(document).ready( function()
   {  
      drawQMEAN('01')
   }) ;

function drawQMEAN(id)
{
  var resCount = 0;
  var local_scores = [];
  for(chain in modelData[id].qmean_local)
  {
    var ls = {'chain':chain,'scores':[]};

    for(var i=0; i<modelData[id].qmean_local[chain].length; i++)
    {
        ls['scores'].push([i, modelData[id].qmean_local[chain][i]]);
        if(modelData[id].qmean_local[chain][i])
          resCount++;
    }
    local_scores.push(ls);
  }

  $('#avgLocalScore_'+id)
    .css({'background-color':
          qmeanLocalColour(modelData[id].qmean_global.avg_local_score, .9),
          'padding':'2px 8px 2px 8px',
          'color':'white',
          'border-radius':'1em',               
          'display':'inline-block'});

  drawQMEANLocal(local_scores, modelData[id].qmean_global.avg_local_score, id);

  drawQMEANGlobal(modelData[id]['qmean_global'], resCount, id, false);

  $('a[rel=lq-'+id+'],a[href="#qn-chart-'+id+'"]')
    .fancybox(
      {
        padding:30, 
        'cyclic':true,
        'hideOnContentClick':true,
        beforeShow:function(a,b)
        {
          var id = $(this.content[0]).first().attr('id');
          if(id.indexOf('lq-chart')>-1)
            $('#'+id).closest('.fancybox-outer').after(
               '<a style="color:#666;float:right;" onclick="svgToPNG(\''+id+'\',\'Local_quality_estimate.png\');">&#128247; Save as PNG</a>'
               );
          else if(id.indexOf('qn-chart')>-1)
            $('#'+id).after(
               '<a style="color:#666;float:right;" onclick="svgToPNG(\''+id+'\',\'Quality_comparison.png\');">&#128247; Save as PNG</a>'
               );
        }
      });
}




  </script>

  </body>
</html>
